cache/in-memory/src/rule.rs

```diff
@@ -0,0 +1,168 @@
+
+use crate::InMemoryCache;
+use std::fmt::Debug;
+use twilight_model::{
+    channel::Channel,
+    guild::{Emoji, GuildIntegration, Member, Role},
+    user::User,
+    voice::VoiceState,
+};
+
+/// Entity being checked on whether it should be accepted into the cache.
+#[derive(Clone, Copy, Debug)]
+#[non_exhaustive]
+pub enum Entity<'a> {
+    /// Channel is being resolved.
+    Channel(&'a Channel),
+    /// Emoji is being resolved.
+    Emoji(&'a Emoji),
+    /// Integration is being resolved.
+    Integration(&'a GuildIntegration),
+    /// Member is being resolved.
+    Member(&'a Member),
+    /// Role is being resolved.
+    Role(&'a Role),
+    /// User is being resolved.
+    User(&'a User),
+    /// Voice state is being resolved.
+    VoiceState(&'a VoiceState),
+}
+
+/// Resolution of a [`Rule`].
+#[derive(Debug)]
+#[non_exhaustive]
+pub enum RuleResolution {
+    /// Accept the entity.
+    Accept,
+    /// Reject the entity.
+    Reject,
+    /// Reject the entity, but only if no other rule has accepted the entity.
+    UnsettledReject,
+}
+
+impl RuleResolution {
+    /// An accepted resolution.
+    pub const ACCEPT: Self = Self::Accept;
+
+    /// A denied resolution.
+    pub const REJECT: Self = Self::Reject;
+
+    /// An unsettled denial resolution.
+    ///
+    /// Refer to [`RuleResolution::UnsettledReject`] for more documentation.
+    pub const UNSETTLED_REJECT: Self = Self::UnsettledReject;
+
+    /// Create a resolution from a boolean.
+    ///
+    /// True maps to [`Self::ACCEPT`] while false maps to [`Self::REJECT`]. This
+    /// does not map to [`Self::UNSETTLED_REJECT`].
+    ///
+    /// # Examples
+    ///
+    /// Create rules from booleans:
+    ///
+    /// ```
+    /// use twilight_cache_inmemory::rule::RuleResolution;
+    ///
+    /// assert_eq!(RuleResolution::ACCEPT, RuleResolution::from_bool(true));
+    /// assert_eq!(RuleResolution::REJECT, RuleResolution::from_bool(true));
+    /// ```
+    pub const fn from_bool(accept: bool) -> Self {
+        if accept {
+            Self::ACCEPT
+        } else {
+            Self::REJECT
+        }
+    }
+}
+
+/// Rules for accepting or rejecting entities into the cache.
+///
+/// The concept of a rule is simple at its heart: given an object, do I want it
+/// to be cached? Rules are predicates that determine this. The cache is
+/// provided to help aid in computing this.
+///
+/// Rules can be added to the cache via [`InMemoryCacheBuilder::rules_mut`].
+///
+/// # Examples
+///
+/// Define a rule to accept only member objects into the cache that are
+/// associated with the current user, rejecting all other member objects:
+///
+/// ```
+/// use twilight_cache_inmemory::{config::{Entity, Rule}, InMemoryCache};
+///
+/// #[derive(Clone, Debug, Eq, PartialEq)]
+/// pub struct CurrentMemberRule;
+///
+/// impl Rule for CurrentMemberRule {
+///     fn accept(&self, cache: &InMemoryCache, entity: &Entity) -> bool {
```

Comment:
Can this documentation be brought in line with itself? Some parts use the enum value, while others use the associated constant of the enum (tbh we might not even need the associated constant)